doc/design/api.md

```diff
@@ -0,0 +1,90 @@
+# Design Doc: PaddlePaddle API
+
+## Ingredients
+
+As the first step of our design, we list important concepts in deep
+learning and try to figure their relationship, as shown below:
+
+```
+Model = {topology, parameters}
+
+Evaluator = {Model*, activations}
+- forward
+- test
+
+GradientMachine = {Model*, gradients}
+- backward
+
+Optimizer = {Model*, Evaluator*, GradientMachine*}
+- train
+- update
+- checkpoint
```

Comment:
Does this checkpoint method mean save everything (model topology, parameters, optimizer states)?

I think checkpoint should be called internally from train, and not exposed.
Instead, expose
```
serialized_state = sgd.save() # save only internal state of optimizer
sgd = paddle.optimizer.SGD(serialized_state) # create SGD optimizer from serialized_state
```
In this way everything is orthogonal (before `optimizer.checkpoint` includes `model.save`), user will have option to save and load whatever they want (typically when they manually implement `train`). By doing:
```
# save
model.save(path)
sgd_state = sgd.save()

# load
model.load(path)
sgd = paddle.optimizer.SGD(sgd_state)
```